### ISSUE TITLE
Composing of functions and funs

### ISSUE THREAD

**Issue by dlfivefifty** [JuliaApproximation/ApproxFun.jl#532]
Code like `x->abs(f(x))` in `abs(f::Fun)` doesn't benefit from the fact that `f` can be evaluated fast at the grid. 

Perhaps we need to override `∘(f, ::Fun)` to return `ComposeFun`, that knows that it can evaluate fast on `ChebyshevPoints`.

**Comment by MikaelSlevinsky** [comment]
We'd need to determine the necessary number of Chebyshev points. For `sign`, it comes from the ratio of minimal root separation & location to domain length (just need one point in between each root), but it would be more involved for `abs`, since it still needs to resolve a function in between roots.

**Comment by dlfivefifty** [comment]
You misunderstand:
```julia
struct ComposedFun{OP,FUN} <: Callable
  op::OP
  fun::FUN
end

∘(op, f::Fun) = ComposedFun(op, f)

(c::ComposedFun)(x) = c.op(c.fun(x))
```

This avoids compilation of `x->abs(f(x))` for each `f`, as it would only compile once for `ComposedFun{typeof(abs), Fun{Chebyshev{Segment{Float64},Float64},Float64,Vector{Float64}}`

Also, when grids are setup, we can do things like have 
```julia
c = abs ∘ f
vals = c.(ChebyshevGrid(10_000))
```
in O(n log n) by overriding
```julia
function broadcast(c::ComposedFun{OP,FUN}, x::ChebyshevGrid) where {OP, FUN <: Fun{<:Chebyshev}} 
   vals = c.fun.(x)    # uses transform to evaluate at the grid
   vals .= c.op.(x)    # then  evaluates op(f(x)), without reallocating
end
```

**Comment by dlfivefifty** [comment]
This way `Fun(abs ∘ f)` automatically becomes an O(n log n) algorithm (instead O(m n + n log n) when `f` needs `m` coefficients)

**Comment by dlfivefifty** [comment]
Also: much more accurate.

**Comment by MikaelSlevinsky** [comment]
That's interesting (and I like it!), but I'll mention that it's a philosophical break from "every function is represented as a `Fun`", since now the user and/or developer decides when to construct a `Fun` and when to compose.

Also, that broadcast is weird because `vals` is not type-stable, but this is yet to be implemented.

When we currently create `abs(f::Fun)`, the roots are found and a `Fun` is created on each subinterval. I understand that `Fun(abs ∘ f)` can evaluate fast on a Chebyshev grid, but how does `Fun(abs ∘ f)` change the fact that the output is a `Fun` on a `UnionDomain` of subintervals, still costing O(mn) since the Chebyshev grids are mapped?

**Comment by dlfivefifty** [comment]
`ComposedFun` is not for users, only for internel use. `abs(f::Fun)` would work exactly as it does now, just instead of calling (in the case of no roots) `Fun(x -> abs(f(x)), d)` we would call `Fun(abs ∘ f, d)`, which then would get the more optimal complexity & accuracy.

**Comment by MikaelSlevinsky** [comment]
Ah, no roots!

**Comment by MikaelSlevinsky** [comment]
It would have to be a special internal function anyway that asserts a sign, since roots is O(m^2)

**Comment by dlfivefifty** [comment]
I thought `roots` is `O(m)`, since it subdivides?

**Comment by dlfivefifty** [comment]
Nevermind, it looks like roots is actually O(m^3) (which seems slow...)

**Comment by MikaelSlevinsky** [comment]
Subdivision results in O(m^2), but algorithms such as FastPolynomialRoots.jl for symmetric (unitary) plus-rank-one are provably fast & backward stable.

I bet that an FMM-accelerated `split_points_on_an_interval` could attain O(m) for simple roots: (1) the resolution power of the basis dictates that oversampling by a constant factor (pad+transform) cannot result in roots in between sufficiently close samples, (2) fast eval via barycentric formula (i.e. [Cauchy matrix-vector multiplication](https://github.com/MikaelSlevinsky/HierarchicalMatrices.jl/blob/master/examples/Cauchy.jl)) + secant/Newton method converges to the only possible roots on the interval (starting guesses determined by sign changes in function samples), (3) keep everything real, (4) there can be at most m roots on the interval. Though there are subtleties, such as multiple roots, but those introduce ill-conditioning into the problem, independent of the algorithm:
```julia
julia> using ApproxFun

julia> f = Fun(x->x^11)
Fun(Chebyshev(【-1.0,1.0】),[0.0, 0.451172, 0.0, 0.322266, 0.0, 0.161133, 0.0, 0.0537109, 0.0, 0.0107422, 0.0, 0.000976563])

julia> roots(f)
1-element Array{Float64,1}:
 0.0331512

julia> complexroots(f)
11-element Array{Complex{Float64},1}:
  -0.0299524+0.00842597im
  -0.0299524-0.00842597im
  -0.0213607+0.0232159im 
  -0.0213607-0.0232159im 
 -0.00570085+0.0317329im 
 -0.00570085-0.0317329im 
   0.0128358+0.0302013im 
   0.0128358-0.0302013im 
   0.0276023+0.0182487im 
   0.0276023-0.0182487im 
   0.0331517+0.0im       
```

**Comment by MikaelSlevinsky** [comment]
Does subdivision + fast rank-structured Hessenberg eigensolvers = O(m)? @ajt60gaibb

**Comment by MikaelSlevinsky** [comment]
EDIT above: subdivision + fast eval via FMM + fast rank-structured Hessenberg eigensolvers = O(m)?

**Comment by dlfivefifty** [comment]
The eigensolver doesn't play a role because the subdivision uses a fixed size, so that's just a constant.

**Comment by ajt60gaibb** [comment]
Fast subdivision, i.e., fast evaluation via FMM or NUDCT, on its own results in O(m).

**Comment by MikaelSlevinsky** [comment]
Right, I remember now that you use `nudct` in chebfun, and there are is an issue with fast evaluation too close to the endpoints.

**Comment by MikaelSlevinsky** [comment]
Barycentric evaluation is forward stable, so I suppose FMM would be fine?

**Comment by ajt60gaibb** [comment]
Here's the discussion:   https://github.com/chebfun/chebfun/pull/2173

**Comment by ajt60gaibb** [comment]
I fixed that endpoint issue, I believe.  But, I've forgotten what I did...

**Comment by dlfivefifty** [comment]
can't we just use angle variables? This would be automatic with the planned ` ChebyshevGrid `

**Comment by dlfivefifty** [comment]
Assuming you can map from a sub-grid back to angles without using ` acos(... cos(t) ...) ` that is do the scaling in angle space. Should just be polynomial or something

**Comment by MikaelSlevinsky** [comment]
I think there are (at least theoretical) [issues](https://www.mcs.anl.gov/~aaustin/pubs/AX2017.pdf) with stability of the periodic barycentric formula, so perhaps x-space is preferable for FMM.

**Comment by dlfivefifty** [comment]
What about NUDCT?

**Comment by dlfivefifty** [comment]
Or do it in coefficient space: it is polynomial to polynomial so there exists a triangular conversion operator. Maybe it's Toeplitz dot Henkel?

**Comment by ajt60gaibb** [comment]
We work in the angle variable for the NUDCT.  Interesting thought to do it in coefficient space...

**Comment by MikaelSlevinsky** [comment]
Maybe it's FMM-able in coefficient-space too! (I realize that's my go-to phrase these days...)

**Comment by MikaelSlevinsky** [comment]
Are the entries of the restriction operator known?

**Comment by dlfivefifty** [comment]
Well they are just integrals. Look kind of like Bessel functions

**Comment by dlfivefifty** [comment]
Actually, from my paper with @marcusdavidwebb we know the conversion operators always satisfy a 5-point recurrence relationship built from the recurrences, so we can at least construct it (but in O(n^2)). Maybe since the Jacobi parameters α_k and β_k are all constant (both Chebyshev), it all simplifies?

**Comment by MikaelSlevinsky** [comment]
sounds like binomials with geometric scaling proportional to restriction width

**Comment by dlfivefifty** [comment]
Yep that sounds likely. If you remember when you visited Sydney we looked at the restriction of Taylor to Taylor and decided it was doable

**Comment by MikaelSlevinsky** [comment]
Even if it's exactly square, it might be the case that the severe scaling makes it numerically rectangular, since the "numerical degree" is related to the restriction length.

**Comment by dlfivefifty** [comment]
That can't be since the restriction of `cos(1_000_000x)` still needs a lot of coefficients to resolve.

**Comment by MikaelSlevinsky** [comment]
You're right, it also depends on the coefficients. Same thing with random O(1) coefficients: such functions are entire yet appear irregular at the same time! If the coefficients decay geometrically or high-order algebraically, maybe more savings are possible.

**Comment by dlfivefifty** [comment]
Just check
```julia
n = 100
    C = zeros(n,n)
    for k = 1:n
        C[1:k,k] = Fun(x->Fun(Chebyshev(),[zeros(k-1);1])(x),Chebyshev(-0.5 .. 0.5),k).coefficients
    end
```
It's numerically less than upper triangular, but also definitely not numerically rectangle (that is the diagonals decay exponentially, but it takes longer to see the decay the larger the band).

**Comment by dlfivefifty** [comment]
(Don't remember how to do `matshow` in Plots.jl so did a contour plkot)

<img width="774" alt="untitled 2" src="https://user-images.githubusercontent.com/1229737/30226644-b0cc2750-94ce-11e7-821d-5365c67658f1.png">

**Comment by MikaelSlevinsky** [comment]
My point was that the truncation depends on the restriction width. Take a look at the matrix with `Chebyshev(-0.125 .. 0.125)`. Of course, this is distinct from rootfinding now because root finding would need `Chebyshev(-1.0 .. 0.0)`, which is worse than above. For `matshow`, I use:
```julia
function logheatmap(A::AbstractMatrix, ϵ)
    LGABSA = [isinf(log10(abs(A[i,j]))) ? log10(ϵ) : max(log10(abs(A[i,j])),log10(ϵ)) for i=1:size(A,2),j=1:size(A,1)]
    heatmap(LGABSA)
end
```

**Comment by MikaelSlevinsky** [comment]
@dlfivefifty, as for the NUDCT, it remains to be implemented. Alex has the most experience and expertise with this.

If we want to cut down on low-degree overhead:

- we could use an in-place `eigvals!` here: https://github.com/JuliaApproximation/ApproxFun.jl/blob/development/src/Extras/roots.jl#L173;

- we could modify the recursive function: https://github.com/JuliaApproximation/ApproxFun.jl/blob/development/src/Extras/roots.jl#L191 to use only two (or even one) arrays; or,

- we could use native Julia eigensolvers (for *very* low degrees): https://github.com/andreasnoack/LinearAlgebra.jl/blob/master/src/eigenGeneral.jl
